src/network/networkd-address.c

```diff
@@ -20,6 +21,85 @@
+int generate_ipv6_ula_address(Link *link, uint16_t subnet_id, Address **ret) {
+        _cleanup_(address_freep) Address *address = NULL;
+        _cleanup_(khash_unrefp) khash *h = NULL;
+        uint8_t key[16], eui64[8], global_id[5];
+        const void *p;
+        uint64_t ntp;
+        int r;
+
+        r = address_new(&address);
+        if (r < 0)
+                return log_link_error_errno(link, r, "Could not allocate address: %m");
+
+        address->family = AF_INET6;
+
+        /* 1. Obtain the current time of day in 64-bit NTP format [NTP]. */
+        ntp = (int64_t) now(CLOCK_MONOTONIC);
+        memcpy((uint8_t *) &key, &ntp, sizeof(uint64_t));
+
+        /* 2) Get EUI-64 from the interface specified */
+        eui64[0]  = link->mac.ether_addr_octet[0];
+        eui64[0] ^= 1 << 1;
+        eui64[1]  = link->mac.ether_addr_octet[1];
+        eui64[2]  = link->mac.ether_addr_octet[2];
+        eui64[3]  = 0xff;
+        eui64[4]  = 0xfe;
+        eui64[5]  = link->mac.ether_addr_octet[3];
+        eui64[6]  = link->mac.ether_addr_octet[4];
+        eui64[7]  = link->mac.ether_addr_octet[5];
+
+        /*  3) Concatenate the time of day with the system-specific identifier
+         *   in order to create a key.
+         */
+        memcpy((uint8_t *) &key + sizeof(uint64_t), (void *) &eui64, sizeof(eui64));
+
+        /* 4) Compute an SHA-1 digest on the key as specified in [FIPS, SHA1];
+         *  the resulting value is 160 bits.
+         */
+
+        r = khash_new(&h, "sha1");
+        if (r < 0)
+                return r;
+
+        r = khash_put(h, &key, sizeof(key));
+        if (r < 0)
+                return r;
+
+        r = khash_digest_data(h, &p);
+        if (r < 0)
+                return r;
+
+        /* 5) Use the least significant 40 bits as the Global ID. Not exactly, just chop off. */
+        memcpy(&global_id, p, MIN(khash_get_size(h), sizeof(global_id)));
+
+        /* 6) Concatenate FC00::/7, the L bit set to 1, and the 40-bit Global
+         *    ID to create a Local IPv6 address prefix.
+         */
+        r = in_addr_prefix_from_string("fc00::/7", AF_INET6, &address->in_addr, &address->prefixlen);
+        if (r < 0)
+                return r;
+
+        /* 3.1.  Format
+         *      The Local IPv6 addresses are created using a pseudo-randomly
+         *       allocated global ID.  They have the following format:
+         *
+         *       | 7 bits |1|  40 bits   |  16 bits  |          64 bits           |
+         *       +--------+-+------------+-----------+----------------------------+
+         *       | Prefix |L| Global ID  | Subnet ID |        Interface ID        |
+         *       +--------+-+------------+-----------+----------------------------+
+         */
+        address->in_addr.in6.s6_addr[0] |= 0x01;
+        memcpy(&address->in_addr.in6.s6_addr[1], &global_id, sizeof(global_id));
+        memcpy(&address->in_addr.in6.s6_addr[7], &subnet_id, sizeof(subnet_id));
```

Comment:
What about the interface ID?

Reply:
I am not sure about it . https://cd34.com/rfc4193/  . Do we need to append it ? @bsjones . remaining 64 bit we need to fill up . Is that EUI64 or something else ?

Reply:
https://tools.ietf.org/rfc/rfc3513.txt It seems EUI64.

Reply:
OK let's make it.